ccl/changefeedccl: TestChangefeedEnvelope failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/31ac673c12094e27a87ae4006efc1591bc8795e1

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestChangefeedEnvelope PKG=github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1095751&tab=buildLog

```

I190116 06:09:55.848984 2277 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
1      [async] changefeed-poller
I190116 06:09:55.856003 2277 util/stop/stopper.go:536  quiescing; tasks left:
1      [async] changefeed-poller
panic: 1 leaked iterators [recovered]
	panic: 1 leaked iterators [recovered]
	panic: 1 leaked iterators

goroutine 2277 [running]:
testing.tRunner.func1(0xc00152cb00)
	/usr/local/go/src/testing/testing.go:792 +0x387
panic(0x2d25340, 0xc0004320b0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000dbbc20, 0x385ca80, 0xc000050108)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:183 +0x11f
panic(0x2d25340, 0xc0004320b0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage/engine.(*RocksDB).Close(0xc0014c0480)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/engine/rocksdb.go:717 +0x30f
github.com/cockroachdb/cockroach/pkg/server.(*Engines).Close(0xc0019f9ab8)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/config.go:400 +0x62
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc000dbbc20, 0x385ca80, 0xc000050108)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:487 +0x366
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.sinklessTest.func1(0xc00152cb00)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:882 +0x37a
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.rangefeedTest.func1(0xc00152cb00)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:932 +0x85
testing.tRunner(0xc00152cb00, 0xc0004078a0)
	/usr/local/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x35c

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0015f0200, 0x30e9d13, 0x16, 0x3212fc0, 0x9d0801)
	/usr/local/go/src/testing/testing.go:879 +0x383
testing.runTests.func1(0xc0001c5900)
	/usr/local/go/src/testing/testing.go:1119 +0x78
testing.tRunner(0xc0001c5900, 0xc00082bde8)
	/usr/local/go/src/testing/testing.go:827 +0xbf
testing.runTests(0xc0004d4940, 0x539d8e0, 0x2c, 0x2c, 0x38f2f98)
	/usr/local/go/src/testing/testing.go:1117 +0x2aa
testing.(*M).Run(0xc0004ffa00, 0x0)
	/usr/local/go/src/testing/testing.go:1034 +0x165
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestMain(0xc0004ffa00)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/main_test.go:31 +0xd7
main.main()
	_testmain.go:128 +0x13d

goroutine 6 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x41

goroutine 8 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1211 +0xf2
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:627 +0x11e

goroutine 9 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:634 +0xab
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:628 +0x136

goroutine 12 [select]:
github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view.(*worker).start(0xc000017800)
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:152 +0xdd
created by github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view.init.0
	/go/src/github.com/cockroachdb/cockroach/vendor/go.opencensus.io/stats/view/worker.go:29 +0x57

goroutine 1280 [chan receive]:
testing.(*T).Run(0xc00152cb00, 0x30ce2bb, 0x9, 0xc0004078a0, 0x337f9b01)
	/usr/local/go/src/testing/testing.go:879 +0x383
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestChangefeedEnvelope(0xc0015f0200)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeed_test.go:113 +0x11c
```

--- a/pkg/storage/rangefeed/processor.go
+++ b/pkg/storage/rangefeed/processor.go
@@ -217,6 +217,9 @@ func (p *Processor) Start(stopper *stop.Stopper, rtsIter engine.SimpleIterator) 
 					}
 				}
 				if err := stopper.RunAsyncTask(ctx, "rangefeed: output loop", runOutputLoop); err != nil {
+					if r.catchupIter != nil {
+						r.catchupIter.Close() // clean up
+					}
 					r.disconnect(roachpb.NewError(err))
 					p.reg.Unregister(&r)
 				}
@@ -352,6 +355,9 @@ func (p *Processor) Register(
 	select {
 	case p.regC <- r:
 	case <-p.stoppedC:
+		if catchupIter != nil {
+			catchupIter.Close() // clean up
+		}
 		// errC has a capacity of 1. If it is already full, we don't need to send
 		// another error.
 		select {

